File: DSA/HR/counting_valleys.java
class Result {
    public static int countingValleys(int steps, String path) {
        int currLevel = 0;
        int valleysCount = 0;
        boolean isInAValley = false;
        for (char c : path.toCharArray()) {
            currLevel += c == 'D' ? -1 : 1;
            if (!isInAValley && currLevel == -1) {
                valleysCount++;
                isInAValley = true;
            } else if (isInAValley && currLevel == 0) {
                isInAValley = false;
            }
        }
        return valleysCount;
    }

}

File: DSA/HR/counting_valleys.cpp
#include <string>

using namespace std;

int countingValleys(int steps, string path)
{
    bool possibleValley{false};
    int valleyCount{};
    int height{};

    for (char c : path)
    {
        if (c == 'D')
        {
            if (height == 0)
                possibleValley = true;
            height--;
        }
        else
        {
            height++;
            if (height == 0 && possibleValley)
            {
                valleyCount++;
                possibleValley = false;
            }
        }
    }

    return valleyCount;
}